ot/gmm.py

```diff
@@ -0,0 +1,332 @@
+
+    Parameters
+    ----------
+    x : array-like, shape (n_samples, d)
+        The input samples.
+    m : array-like, shape (d,)
+        The mean vector of the Gaussian distribution.
+    C : array-like, shape (d, d)
+        The covariance matrix of the Gaussian distribution.
+
+    Returns
+    -------
+    pdf : array-like, shape (n_samples,)
+        The probability density function evaluated at each sample.
+
+    """
+    _, d = x.shape
+    z = (2 * np.pi) ** (-d / 2) * np.linalg.det(C) ** (-0.5)
+    exp = np.exp(-0.5 * np.sum((x - m) @ np.linalg.inv(C) * (x - m), axis=1))
+    return z * exp
+
+
+def gmm_pdf(x, m, C, w):
+    r"""
+    Compute the probability density function (PDF) of a Gaussian Mixture Model (GMM) at given points.
+
+    Parameters
+    ----------
+    x : array-like, shape (n_samples, d)
+        The input samples.
+    m : array-like, shape (n_components, d)
+        The means of the Gaussian components.
+    C : array-like, shape (n_components, d, d)
+        The covariance matrices of the Gaussian components.
+    w : array-like, shape (n_components,)
+        The weights of the Gaussian components.
+
+    Returns
+    -------
+    out : array-like, shape (n_components,)
+        The PDF values at the given points.
+
+    """
+    out = np.zeros((x.shape[0]))
+    for k in range(m.shape[0]):
+        out = out + w[k] * gaussian_pdf(x, m[k], C[k])
+    return out
+
+
+def dist_bures_squared(m_s, m_t, C_s, C_t):
+    r"""
+    Compute the matrix of the squared Bures distances between the components of
+    two Gaussian Mixture Models (GMMs). Used to compute the GMM Optimal
+    Transport distance [69].
+
+    Parameters
+    ----------
+    m_s : array-like, shape (k_s, d)
+        Mean vectors of the source GMM.
+    m_t : array-like, shape (k_t, d)
+        Mean vectors of the target GMM.
+    C_s : array-like, shape (k_s, d, d)
+        Covariance matrices of the source GMM.
+    C_t : array-like, shape (k_t, d, d)
+        Covariance matrices of the target GMM.
+
+    Returns
+    -------
+    dist : array-like, shape (k_s, k_t)
+        Matrix of squared Bures distances between the components of the source
+        and target GMMs.
+
+    References
+    ----------
+    .. [69] Delon, J., & Desolneux, A. (2020). A Wasserstein-type distance in
+    the space of Gaussian mixture models. SIAM Journal on Imaging Sciences,
+    13(2), 936-970.
+
+    """
+    nx = get_backend(m_s, C_s, m_t, C_t)
+    k_s, k_t = m_s.shape[0], m_t.shape[0]
+
+    assert m_s.shape[0] == C_s.shape[0], \
+        "Source GMM has different amount of components"
+
+    assert m_t.shape[0] == C_t.shape[0], \
+        "Target GMM has different amount of components"
+
+    assert m_s.shape[-1] == m_t.shape[-1] == C_s.shape[-1] == C_t.shape[-1], \
+        "All GMMs must have the same dimension"
+
+    D_means = dist(m_s, m_t, metric='sqeuclidean')
+    D_covs = nx.zeros((k_s, k_t), type_as=m_s)
+
+    for i in range(k_s):
```

Comment:
Challenge (optional): implement eig and sqrtm that can hadle batches of matrices and do the whole matrix with a magnificent einsum !

Reply:
done! :D

Reply:
done :D